target/snitch_cluster/sw/snax/streamer-simd/src/snax-streamer-simd-lib.c

```diff
@@ -0,0 +1,162 @@
+    // temporal loop bounds, from innermost to outermost
+    write_csr(960 + 0, tempLoop0);
+    write_csr(960 + 1, tempLoop1);
+
+    // temporal strides for data reader (In)
+    write_csr(960 + 2, tempStride0_in);
+    write_csr(960 + 3, tempStride1_in);
+
+    // temporal strides for data writer (Out)
+    write_csr(960 + 4, tempStride0_out);
+    write_csr(960 + 5, tempStride1_out);
+
+    // fixed spatial strides for data reader (In)
+    write_csr(960 + 6, 4);
+    write_csr(960 + 7, 32);
+
+    // fixed spatial strides for data writer (Out)
+    write_csr(960 + 8, 1);
+    write_csr(960 + 9, 8);
+
+    // base ptr for data reader (In)
+    write_csr(960 + 10, (uint32_t)(delta_local_in + snrt_l1_next()));
+
+    // base ptr for data writer (Out)
+    write_csr(960 + 11, (uint32_t)(delta_local_out + snrt_l1_next()));
+}
+
+void start_streamer_simd() { write_csr(960 + 12, 1); }
+
+void set_simd_csr(uint32_t csr0, uint32_t csr1, uint32_t csr2,
+                  uint32_t temporal_loop_bound) {
+    // set the constants for the SIMD unit
+    write_csr(960 + 14, csr0);
+    write_csr(960 + 15, csr1);
+    write_csr(960 + 16, csr2);
+
+    // set the temporal loop bound
+    write_csr(960 + 17, temporal_loop_bound);
+}
+
+void start_simd() { write_csr(960 + 18, 1); }
+
+void wait_streamer_simd() {
+    write_csr(960 + 18, 0);
+    write_csr(960 + 18, 0);
+    write_csr(960 + 12, 0);
+}
+
+uint32_t read_simd_streamer_perf_counter() {
+    uint32_t perf_counter = read_csr(960 + 13);
+    return perf_counter;
+}
+
+uint32_t read_simd_state() {
+    uint32_t status = read_csr(960 + 19);
+    return status;
+}
+
+uint32_t read_simd_perf_counter() {
+    uint32_t perf_counter = read_csr(960 + 20);
+    return perf_counter;
+}
+
+void load_simd_test_data(int tempLoop0, int tempLoop1, int tempStride0,
+                         int tempStride1, int32_t* base_ptr_local,
+                         int32_t* base_ptr_l2) {
+    int32_t* addr_in;
+    int32_t* addr_In;
+
+    for (int loop1 = 0; loop1 < tempLoop1; loop1++) {
+        for (int loop0 = 0; loop0 < tempLoop0; loop0++) {
+            addr_in =
+                base_ptr_local +
+                (loop1 * tempStride1 + loop0 * tempStride0) / sizeof(int32_t);
+            addr_In =
+                base_ptr_l2 + loop1 * tempLoop0 * vec_len + loop0 * vec_len;
+            snrt_dma_start_1d(addr_in, addr_In, vec_len * sizeof(int32_t));
+        }
+    }
+}
+
+int8_t scale_quant_clamp_c_spec(int32_t input, int8_t input_zp,
+                                int8_t output_zp, int32_t multiplier,
+                                int8_t shift,  // values between 0-63
+                                int8_t max_int, int8_t min_int,
+                                bool double_round) {
+    // input zero-point adjustment
+    input = input - input_zp;
+
+    // multiplication
+    int64_t var0 = (int64_t)input * (int64_t)multiplier;
+
+    // shift & round
+    int32_t var1 = var0 >> (shift - 1);
```

Comment:
What if shift is 0?

Reply:
Discussed offline! Please refer to: https://gist.github.com/jorendumoulin/83352a1e84501ec4a7b3790461fee2bf